src/components/VmActions/AutoSelectConsole.js

```diff
@@ -0,0 +1,133 @@
+import React, { useEffect, useState } from 'react'
+import { connect } from 'react-redux'
+import PropTypes from 'prop-types'
+import { addUserMessage, downloadVmConsoles } from '_/actions'
+import ConsoleConfirmationModal from '_/components/VmActions/ConsoleConfirmationModal'
+import { msg } from '_/intl'
+
+const AutoSelectConsole = (props) => {
+  const {
+    vm,
+    show,
+    defaultConsole,
+    isNoVNC,
+    fetchConsoles,
+    requestActive,
+    id,
+    onFailure,
+  } = props
+
+  const initialState = {
+    // consoles: [],
+    selectedProtocol: undefined,
+    failureReason: undefined,
+    dispatchStatus: 'doNotDispatch',
+    active: false,
+  }
+
+  const [state, setState] = useState(initialState)
+  const {
+    dispatchStatus,
+    selectedProtocol,
+    active,
+    // failureReason,
+    // consoles,
+  } = state
+
+  useEffect(() => {
+    if (show) {
+      setState(state => ({
+        ...state,
+        dispatchStatus: 'shouldDispatch',
+        active: true,
+      }))
+      fetchConsoles(vm.get('id'))
+    }
+  }, [show]) // eslint-disable-line react-hooks/exhaustive-deps
+
+  useEffect(() => {
+    const newState = {}
+
+    const { selectedProtocol, failureReason } = selectProtocol(vm.get('consoles'), requestActive)
+
+    if (selectedProtocol || failureReason) {
+      newState.selectedProtocol = selectedProtocol
+      newState.failureReason = failureReason
+    }
+    if (active && selectedProtocol && dispatchStatus !== 'doNotDispatch') newState.dispatchStatus = 'dispatched' // cannot change from 'doNotDispatch' to 'dispatched'
+    if (active && !selectedProtocol && failureReason && !requestActive && dispatchStatus !== 'doNotDispatch') onFailure(failureReason)
+
+    setState(state => ({ ...state, ...newState }))
+  }, [vm, requestActive]) // eslint-disable-line react-hooks/exhaustive-deps
+
+  const onClose = () => {
+    if (dispatchStatus === 'dispatched') setState(state => ({ ...state, dispatchStatus: 'doNotDispatch', active: false })) // can only change from 'dispatched' to 'doNotDispatch'
+    props.onClose()
+  }
+
+  const selectProtocol = (consoles, requestActive) => {
+    let selectedProtocol, failureReason
+    if (!requestActive) {
+      if (consoles.size > 0) {
+        if (consoles.size === 1) {
+          selectedProtocol = consoles.first()
+        } else {
+          selectedProtocol = consoles.find(c => c.get('protocol') === defaultConsole)
+          failureReason = !selectedProtocol && 'noDefault'
+        }
+      } else {
+        failureReason = 'emptyConsolesList'
+      }
+    }
+    return { selectedProtocol, failureReason }
```

Comment:
Function can be made pure by moving `defaultConsole` to parameters. We could then move it outside the component and unit test. This would also allow to use it directly as hook dependency.

Reply:
Done.